src/internal/modules/address-entry/controller.js

```diff
@@ -0,0 +1,98 @@
+const forms = require('shared/lib/forms');
+const newAddress = require('../../lib/new-address');
+const helpers = require('./lib/helpers');
+const { omit } = require('lodash');
+
+const sessionForms = require('shared/lib/session-forms');
+
+const storeAddressAndRedirect = (request, h, address) => {
+  newAddress.set(request, address);
+  const redirectPath = helpers.getRedirectPath(request);
+  return h.redirect(redirectPath);
+};
+
+const getPostcode = (request, h) => {
+  helpers.saveReferenceData(request);
+  return h.view('nunjucks/address-entry/enter-uk-postcode', {
+    ...request.view,
+    ...helpers.getPageCaption(request),
+    pageTitle: 'Enter the UK postcode',
+    back: request.query.back,
+    form: sessionForms.get(request, ukPostcode.form(request))
+  });
+};
+
+const postPostcode = async (request, h) => {
+  const { postcode } = request.payload;
+  const form = forms.handleRequest(
+    ukPostcode.form(request, postcode),
+    request,
+    ukPostcode.schema
+  );
+
+  if (form.isValid) {
+    return h.redirect(helpers.getSelectAddressUrl(request));
+  }
+  const { redirectPath, back } = request.yar.get(helpers.SESSION_KEY);
+  return h.postRedirectGet(form, '/address-entry/postcode', { redirectPath, back });
+};
+
+const getSelectAddress = (request, h) => {
+  helpers.setAddressSearchResults(request);
+  return h.view('nunjucks/address-entry/select-address', {
+    ...request.view,
+    ...helpers.getPageCaption(request),
+    pageTitle: 'Select the address',
+    back: helpers.getPostcodeUrl(request),
+    postcode: helpers.getPostcode(request),
+    form: sessionForms.get(request, selectAddress.form(request))
+  });
+};
+
+const postSelectAddress = (request, h) => {
+  const { uprn } = request.payload;
+  const form = forms.handleRequest(
+    selectAddress.form(request, uprn),
+    request,
+    selectAddress.schema
+  );
+
+  if (form.isValid) {
+    const addressSearchResults = helpers.getAddressSearchResults(request);
+    const selectedAddress = addressSearchResults.find(address => address.uprn === parseInt(uprn));
+    return storeAddressAndRedirect(request, h, selectedAddress);
+  }
+
+  return h.postRedirectGet(form, '/address-entry/address/select', { postcode: helpers.getPostcode(request) });
+};
+
+const getManualAddressEntry = (request, h) => h.view('nunjucks/form', {
+  ...request.view,
+  ...helpers.getPageCaption(request),
+  pageTitle: 'Enter the address',
+  back: helpers.getManualAddressEntryBackLink(request),
+  form: sessionForms.get(request, manualAddressEntry.form(request))
+});
+
+const postManualAddressEntry = (request, h) => {
+  let form = forms.handleRequest(
+    manualAddressEntry.form(request, request.payload),
+    request,
+    manualAddressEntry.schema
+  );
+
+  form = manualAddressEntry.applyRequiredFieldErrors(form, request.payload);
+
+  if (form.isValid) {
+    return storeAddressAndRedirect(request, h, omit(request.payload, ['csrfToken']));
```

Comment:
I assume this would overwrite any uprn already in the session from a previous request to the other form?  
This would be desirable because we wouldn't want them to select an address, then go and edit a custom address so the address didn't match the UPRN.